roscala/src/main/scala/coop/rchain/rosette/prim/tuple.scala

```diff
@@ -118,6 +118,48 @@ object tuple {
     }
   }
 
+  /**
+    * Define the "tuple-xchg" primitive.
+    * This exchanges the nth and mth elements of the specified Tuple
+    * e.g. (tuple-xchg [1 2 3 4 5] 1 3) ==> [1 4 3 2 5]
+    */
+  object tplXchg extends Prim {
+    override val name: String = "tuple-xchg"
+    override val minArgs: Int = 3
+    override val maxArgs: Int = 3
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Tuple] = {
+      val elem = ctxt.argvec.elem
+
+      // Check and get arguments: Tuple, Fixnum, Fixnum
+      checkTuple(0, elem).flatMap(
+        t => // Ensure arg0 is a Tuple
+          checkFixnum(1, elem).flatMap(n => // Ensure arg1 is a Fixnum
+            checkFixnum(2, elem).flatMap { m => // Ensure arg2 is a Fixnum
+
+              t.nth(n.value) match { // Extract nth
+                case Some(nv: Fixnum) =>
+                  t.nth(m.value) match { // Extract mth
+                    case Some(mv: Fixnum) =>
+                      t.setNth(n.value, mv) match { //  set mth
+                        case Some(t1: Tuple) =>
+                          t1.setNth(m.value, nv) match { // Set nth
+                            case Some(ret: Tuple) => Right(ret) // Success
+                            case None =>
+                              Left(IndexOutOfBounds(m.value, t.elem.size)) // Fail set nth
+                          }
+                        case None => Left(IndexOutOfBounds(n.value, t.elem.size)) // Fail set mth
+                      }
+                    case None => Left(IndexOutOfBounds(m.value, t.elem.size)) // Fail extract mth
+                  }
+                case None => Left(IndexOutOfBounds(n.value, t.elem.size)) // Fail extract nth
+              }
```

Comment:
You could replace the whole thing with a for-comprehension. You can define something like this:
```scala
def wrapper[A,B](f: Int => Option[Fixnum], v: Int, size: Int): Either[Err,Ob] = {
  f(v).toRight(IndexOutOfBounds(v, size))
}
```
And then use it like:
```scala
val f = wrapper(_, _, t.elem.size)

for {
  v1 <- f(t.nth, n.value)
  v2 <- f(t.nth, m.value)
  v3 <- f(t.setNth(_, mv), n.value)
  v4 <- f(t.setNth(_, nv), m.value)
} yield v4
```

Probably top `flatMap`s could be also placed in `for`.

Reply:
This will take a while to digest.  You're speaking to a 1st year child with college level language.  I'll look into it and try to understand it.